File: apps/ios/App/Screens/EventFormScreen.swift
import SwiftUI
import TrackEvolutionKit

/// New event, or edit an existing one. `viewEventForm` in `public/app.js` is the
/// reference for fields and copy.
///
/// Two rules this screen exists to not break:
///
/// - **The track name carries the layout.** "Virginia International Raceway (Full)"
///   and "(Patriot)" are different circuits that time differently, and the server
///   resolves tracks by name `COLLATE NOCASE`. So the field is passed through with
///   nothing but a whitespace trim: no title-casing, no punctuation tidying, no
///   collapsing of the parenthetical. Normalizing here silently merges two layouts'
///   personal bests, and there is no undo for that.
/// - **`car` is free text.** The server auto-matches it to a garage vehicle by name.
///   The garage suggestions are a convenience, not a picker — typing a car that
///   isn't in the garage has to keep working.
///
/// A **new** event ends with an "Add laps" card — the event page's "Add a session"
/// card, minus the recorder: *Import video* opens the same importer, whose review
/// hands its sessions back through `AppRouter.stagedSessions`
/// (`Route.importVideo(forNewEvent:)`), and lap times can be typed straight in.
/// Both are saved with the event, so logging a day is one screen rather than a
/// form and then a page. Editing an existing event has no such card: its page
/// already has the real one.
struct EventFormScreen: View {
    let target: EventFormTarget

    @Environment(AuthController.self) private var auth
    @Environment(AppRouter.self) private var router
    @Environment(\.dismiss) private var dismiss

    @State private var model: EventFormModel?

    var body: some View {
        TELoadable(state: model?.state ?? .loading, retry: { await model?.load() }) {
            if let model {
                form(model)
            }
        }
        .navigationTitle(model?.isEditing == true ? "Edit event" : "New event")
        .navigationBarTitleDisplayMode(.inline)
        .task {
            if model == nil {
                let model = EventFormModel(api: auth.api, target: target, units: auth.units)
                self.model = model
                await model.load()
            }
            takeStagedSessions()
        }
        // The importer pushed from this form pops back onto it with the sessions
        // it staged; the form stays alive underneath a push, so this fires.
        .onChange(of: router.stagedSessions) { _, _ in takeStagedSessions() }
    }

    /// Adopt whatever an import staged for this form, and empty the hand-off.
    private func takeStagedSessions() {
        guard let model, !model.isEditing, !router.stagedSessions.isEmpty else { return }
        model.stage(router.stagedSessions)
        router.stagedSessions = []
    }

    private func form(_ model: EventFormModel) -> some View {
        // Every field is owned by the model, so the bindings come from it: a form this
        // long with @State per field would drift out of step with a re-load.
        @Bindable var model = model
        return TEPage {
            TECard {
                VStack(alignment: .leading, spacing: 16) {
                    TEField(
                        label: "Track",
                        hint: """
                            Pick from your tracks and known US tracks, or type a new name — layouts time \
                            differently, so name them separately ("Virginia International Raceway (Full)" \
                            vs "(Patriot)") to keep PBs honest
                            """
                    ) {
                        SuggestingField(
                            placeholder: "Virginia International Raceway (Full)",
                            text: $model.trackName,
                            options: model.trackOptions
                        )
                    }

                    TEField(label: "Start date") {
                        DatePicker(
                            "Start date",
                            selection: $model.startDate,
                            displayedComponents: .date
                        )
                        .labelsHidden()
                        .datePickerStyle(.compact)
                    }

                    TEField(label: "Days", hint: "Half days count — a Saturday plus a Sunday morning is 1.5") {
                        HStack(spacing: 12) {
                            Text(fmtDays(model.days))
                                .teStyle(.bodyStrong)
                                .foregroundStyle(Color(.textStrong))
                            Stepper("Days", value: $model.days, in: 0.5...30, step: 0.5)
                                .labelsHidden()
                        }
                    }

                    TEField(
                        label: "On-track hours (optional)",
                        hint: "Seat time for consumable wear tracking — leave blank for the 2h-per-day estimate"
                    ) {
                        TextField("est. 2h per day", text: $model.trackHours)
                            .teInput()
                            .keyboardType(.decimalPad)
                    }

                    TEField(label: "Club / organizer") {
                        TextField("VIR Club", text: $model.club)
                            .teInput()
                    }

                    TEField(label: "Run group") {
                        TextField("High Speed", text: $model.runGroup)
                            .teInput()
                    }

                    TEField(label: "Car", hint: "Pick from your garage or type anything") {
                        SuggestingField(
                            placeholder: "Corvette Z06, Miata, GT3…",
                            text: $model.car,
                            options: model.carOptions
                        )
                    }

                    TEField(label: "Conditions") {
                        Picker("Conditions", selection: $model.conditions) {
                            Text("—").tag(Conditions?.none)
                            ForEach(Conditions.all, id: \.rawValue) { condition in
                                Text(condition.label).tag(Optional(condition))
                            }
                        }
                        .pickerStyle(.segmented)
                    }

                    TEField(label: "Temp \(Units.tempUnit(auth.units)) (optional)") {
                        TextField(String(Units.tempInputSpec(auth.units).placeholder), text: $model.temp)
                            .teInput()
                            .keyboardType(.numbersAndPunctuation)
                    }

                    TEField(
                        label: "Best time (optional)",
                        hint: "Only needed when you don't log laps — logged laps compute this automatically"
                    ) {
                        TextField("2:01.24", text: $model.bestTime)
                            .teInput()
                            .keyboardType(.numbersAndPunctuation)
                            .autocorrectionDisabled()
                    }

                    TEField(label: "Notes") {
                        TextField("Weather, setup changes, incidents…", text: $model.notes, axis: .vertical)
                            .teInput()
                            .lineLimit(3...6)
                    }
                }
            }

            if !model.isEditing {
                addLapsCard(model)
            }

            if let error = model.error {
                TEErrorBanner(message: error)
            }

            Button(model.isSaving ? "Saving…" : model.submitTitle) {
                Task {
                    guard let id = await model.save() else { return }
                    Haptics.confirm()
                    if model.isEditing {
                        dismiss()
                    } else {
                        // Replace the form in the stack with the event it created:
                        // going "back" from a new event should reach the dashboard,
                        // not the form again.
                        router.path = [.event(id)]
                    }
                }
            }
            .buttonStyle(TEButtonStyle(kind: .accent))
            .disabled(model.isSaving || model.trackName.trimmingCharacters(in: .whitespaces).isEmpty)
        }
        // The keyboard must not cover the field being typed into; a form this long
        // needs the scroll view to dismiss it interactively too.
        .scrollDismissesKeyboard(.interactively)
    }

    // MARK: - Add laps

    /// The sessions the new event is created with: clips staged by the importer,
    /// each removable until Create, and one typed by hand. The copy is the event
    /// page card's, so the two read as the same thing.
    private func addLapsCard(_ model: EventFormModel) -> some View {
        @Bindable var model = model
        return TECard {
            VStack(alignment: .leading, spacing: 16) {
                VStack(alignment: .leading, spacing: 4) {
                    Text("Add laps")
                        .teStyle(.h2)
                        .foregroundStyle(Color(.textStrong))
                    Text("Optional — the sessions here are saved with the event. You can always add more from its page.")
                        .teStyle(.xs)
                        .foregroundStyle(Color(.textMuted))
                }

                VStack(alignment: .leading, spacing: 10) {
                    Text("Import a video")
                        .teStyle(.h3)
                        .foregroundStyle(Color(.textStrong))
                    Text("PDR and GoPro clips carry telemetry. Pick one from Files or Photos and the laps come out of it — the video stays on this phone.")
                        .teStyle(.xs)
                        .foregroundStyle(Color(.textMuted))
                    Button("Import video") {
                        router.push(.importVideo(eventId: nil, incoming: nil, forNewEvent: true))
                    }
                    .buttonStyle(TEButtonStyle(kind: .quiet))
                    .accessibilityIdentifier("formImportEntry")

                    ForEach(Array(model.stagedSessions.enumerated()), id: \.offset) { index, draft in
                        HStack(spacing: 12) {
                            VStack(alignment: .leading, spacing: 2) {
                                Text(draft.label ?? "Imported session")
                                    .teStyle(.sm)
                                    .foregroundStyle(Color(.textStrong))
                                    .lineLimit(1)
                                Text(EventFormSessions.stagedSummary(laps: draft.laps ?? []))
                                    .teStyle(.xs)
                                    .foregroundStyle(Color(.textMuted))
                            }
                            Spacer()
                            Button("Remove") { model.removeStaged(at: index) }
                                .teStyle(.xs)
                                .foregroundStyle(Color(.textMuted))
                        }
                        .accessibilityElement(children: .combine)
                        .accessibilityIdentifier("stagedSession\(index)")
                    }
                }

                Divider().overlay(Color(.borderHairline))

                VStack(alignment: .leading, spacing: 10) {
                    Text("Enter lap times by hand")
                        .teStyle(.h3)
                        .foregroundStyle(Color(.textStrong))
                    TEField(label: "Session label") {
                        TextField("Day 1 — Session 2", text: $model.sessionLabel)
                            .teInput()
                    }
                    TEField(
                        label: "Lap times",
                        hint: "Comma, space or newline separated. Formats: 2:01.24 · 2:01 · 121.24 (seconds)"
                    ) {
                        TextField("2:03.55, 2:01.24, 2:02.61", text: $model.sessionLaps, axis: .vertical)
                            .teInput()
                            .keyboardType(.numbersAndPunctuation)
                            .autocorrectionDisabled()
                            .lineLimit(2...5)
                            .accessibilityIdentifier("formSessionLaps")
                    }
                    TEField(label: "Session notes") {
                        TextField("Traffic, tire pressures, line changes…", text: $model.sessionNotes)
                            .teInput()
                    }
                }
            }
        }
    }
}

/// A text field with suggestions under it — the native counterpart to `bindCombo`.
///
/// Free text with hints, **not** a picker: the track field has to accept a circuit
/// the catalog has never heard of, and the car field a car that isn't in the garage.
struct SuggestingField: View {
    let placeholder: String
    @Binding var text: String
    let options: [String]

    @FocusState private var focused: Bool

    var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            TextField(placeholder, text: $text)
                .teInput()
                .autocorrectionDisabled()
                // Track names are proper nouns with parenthetical layouts; iOS
                // autocapitalization would fight the user on "(Full)" vs "(full)"
                // and the server matches case-insensitively anyway.
                .textInputAutocapitalization(.words)
                .focused($focused)

            if focused, !matches.isEmpty {
                ScrollView(.horizontal, showsIndicators: false) {
                    HStack(spacing: 8) {
                        ForEach(matches, id: \.self) { option in
                            Button(option) {
                                text = option
                                focused = false
                            }
                            .teStyle(.xs)
                            .foregroundStyle(Color(.textBody))
                            .padding(.horizontal, 10)
                            .padding(.vertical, 6)
                            .background(Color(.surfaceRaised), in: .capsule)
                            .overlay(Capsule().strokeBorder(Color(.borderHairline), lineWidth: 1))
                        }
                    }
                    .padding(.vertical, 2)
                }
            }
        }
    }

    /// Substring matching, case-insensitive, capped so the strip stays a strip.
    private var matches: [String] {
        let query = text.trimmingCharacters(in: .whitespaces).lowercased()
        guard !query.isEmpty else { return Array(options.prefix(8)) }
        // An exact hit needs no suggestion — offering the thing already typed is noise.
        if options.count == 1, options[0].lowercased() == query { return [] }
        return Array(options.filter { $0.lowercased().contains(query) }.prefix(8))
    }
}

/// The form's fields and the one write it makes.
@MainActor
@Observable
final class EventFormModel {
    private let api: APIClient
    let target: EventFormTarget

    private(set) var state: LoadState = .loading
    private(set) var isSaving = false
    var error: String?

    // Fields, as strings where the server takes a nullable column: an empty string
    // is "clear it", which a numeric binding can't express.
    var trackName = ""
    var startDate = Date()
    var days: Double = 2
    var trackHours = ""
    var club = ""
    var runGroup = ""
    var car = ""
    var conditions: Conditions?
    /// The temperature as typed, in the account's unit system. Stored as whole °F
    /// (`Units.tempToStored`), so a metric entry never drifts on the next edit.
    var temp = ""
    var bestTime = ""
    var notes = ""

    /// The hand-typed session of the "Add laps" card, new events only.
    var sessionLabel = ""
    var sessionLaps = ""
    var sessionNotes = ""
    /// Sessions the import review staged for this event, in posting order.
    private(set) var stagedSessions: [SessionDraft] = []
    /// The event `save()` already created, when a session post after it failed.
    /// nil until then; the next save skips `POST /events` and posts the sessions
    /// still pending — never creating the event twice.
    private(set) var createdId: Int?

    private(set) var trackOptions: [String] = []
    private(set) var carOptions: [String] = []
    /// The track id the event already has, so an unedited name doesn't re-resolve.
    private var existingTrackId: Int?
    /// The system the temperature field is typed in — the account's when the form
    /// opened. Fixed for the form's life so a pre-fill and its save agree.
    let units: UnitSystem

    init(api: APIClient, target: EventFormTarget, units: UnitSystem) {
        self.units = units
        self.api = api
        self.target = target
    }

    var isEditing: Bool {
        if case .edit = target { return true }
        return false
    }

    /// The submit button: the event exists and a session didn't make it, so the
    /// same button now only retries the laps.
    var submitTitle: String {
        if isEditing { return "Save changes" }
        return createdId == nil ? "Create event" : "Add the laps"
    }

    func stage(_ drafts: [SessionDraft]) {
        stagedSessions += drafts
    }

    func removeStaged(at index: Int) {
        guard stagedSessions.indices.contains(index) else { return }
        stagedSessions.remove(at: index)
    }

    /// Everything the "Add laps" card would post, in posting order — the rule
    /// pinned by `contracts/logic/event-form.json`.
    var pendingSessions: [SessionDraft] {
        EventFormSessions.sessionsToCreate(
            staged: stagedSessions, label: sessionLabel, laps: sessionLaps, notes: sessionNotes
        )
    }

    func load() async {
        do {
            async let tracks = api.tracks()
            async let catalog = api.catalog()
            // The garage is deferred, but `/vehicles` is the car field's suggestion
            // list and the source of the default car on a new event.
            async let vehicles = api.vehicles()
            let loaded = try await (tracks: tracks, catalog: catalog, vehicles: vehicles)

            // Your own tracks first, then the rest of the seeded catalog.
            let own = loaded.tracks.map(\.name)
            let seen = Set(own.map { $0.lowercased() })
            trackOptions = own + loaded.catalog.map(\.name).filter { !seen.contains($0.lowercased()) }
            carOptions = loaded.vehicles.map(\.name)

            switch target {
            case .new(let presetTrack):
                trackName = presetTrack ?? ""
                car = loaded.vehicles.first(where: \.isDefault)?.name ?? ""
            case .edit(let id):
                let detail = try await api.event(id: id)
                fill(from: detail.event)
            }
            state = .ready
        } catch let error as APIError {
            state = .failed(error.message)
        } catch {
            state = .failed(error.localizedDescription)
        }
    }

    private func fill(from event: Event) {
        trackName = event.trackName
        existingTrackId = event.trackId
        startDate = EventDates.date(fromISO: event.startDate) ?? Date()
        days = event.days
        trackHours = event.trackHours.map { fmtNumber($0) } ?? ""
        club = event.club ?? ""
        runGroup = event.runGroup ?? ""
        car = event.car ?? ""
        conditions = event.conditions
        temp = Units.tempToDisplay(event.tempF, units).map(String.init) ?? ""
        bestTime = event.bestTimeMs.map { LapTime.fmtMs($0) } ?? ""
        notes = event.notes ?? ""
    }

    /// The typed temperature as the whole °F the server stores; nil clears it.
    private var storedTemp: Int? {
        Units.tempToStored(Double(temp.trimmingCharacters(in: .whitespaces)), units)
    }

    /// Create or update. Returns the event's id on success.
    func save() async -> Int? {
        error = nil
        // Trim only. See the note on the screen: any further normalization merges
        // layouts that must stay apart.
        let name = trackName.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !name.isEmpty else {
            error = "A track name is required."
            return nil
        }
        let raw = bestTime.trimmingCharacters(in: .whitespaces)
        let best = raw.isEmpty ? nil : LapTime.parseTime(raw)
        if !raw.isEmpty, best == nil {
            error = "Couldn't parse best time \"\(raw)\" — use 2:01.24 format."
            return nil
        }
        let hoursRaw = trackHours.trimmingCharacters(in: .whitespaces)
        if !hoursRaw.isEmpty, Double(hoursRaw) == nil {
            error = "Couldn't parse on-track hours \"\(hoursRaw)\" — use a number like 4 or 4.5."
            return nil
        }

        isSaving = true
        defer { isSaving = false }

        let iso = EventDates.isoString(from: startDate)
        do {
            switch target {
            case .new:
                let id: Int
                if let createdId {
                    id = createdId
                } else {
                    var draft = EventDraft(startDate: iso, trackName: name)
                    draft.days = days
                    draft.trackHours = Double(hoursRaw)
                    draft.club = nilIfEmpty(club)
                    draft.runGroup = nilIfEmpty(runGroup)
                    draft.car = nilIfEmpty(car)
                    draft.conditions = conditions
                    draft.tempF = storedTemp
                    draft.bestTimeMs = best
                    draft.notes = nilIfEmpty(notes)
                    id = try await api.createEvent(draft)
                    createdId = id
                }
                // Then the laps, onto the event that now exists: the staged
                // imports first, then the hand-typed session. Each is posted once
                // — a posted one leaves the staging (or empties the typed laps)
                // before the next, so a failure leaves exactly the rest.
                for session in pendingSessions {
                    _ = try await api.createSession(eventId: id, session)
                    if stagedSessions.first == session {
                        stagedSessions.removeFirst()
                    } else {
                        sessionLaps = ""
                    }
                }
                return id
            case .edit(let id):
                var patch = EventPatch()
                // The name is always sent: the server find-or-creates the track from
                // it, which is also how an event moves between layouts.
                patch.trackName = .set(name)
                patch.startDate = .set(iso)
                patch.days = .set(days)
                patch.trackHours = .set(Double(hoursRaw))
                patch.club = .set(nilIfEmpty(club))
                patch.runGroup = .set(nilIfEmpty(runGroup))
                patch.car = .set(nilIfEmpty(car))
                patch.conditions = .set(conditions)
                patch.tempF = .set(storedTemp)
                patch.bestTimeMs = .set(best)
                patch.notes = .set(nilIfEmpty(notes))
                try await api.updateEvent(id: id, patch)
                return id
            }
        } catch let error as APIError {
            self.error = sessionFailurePrefix + error.message
            return nil
        } catch {
            self.error = sessionFailurePrefix + error.localizedDescription
            return nil
        }
    }

    /// Once the event exists, a failure is a session's, and the message says so.
    private var sessionFailurePrefix: String {
        createdId == nil ? "" : "The event was created, but a session couldn't be added: "
    }

    private func nilIfEmpty(_ text: String) -> String? {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }

    /// "4" not "4.0" — the hours field is typed by hand and read back.
    private func fmtNumber(_ value: Double) -> String {
        value == value.rounded() ? String(Int(value)) : String(value)
    }
}

File: apps/android/app/src/main/kotlin/app/trackevolution/screens/EventFormScreen.kt
package app.trackevolution.screens

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.HorizontalDivider
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.focus.onFocusChanged
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import app.trackevolution.core.model.Conditions
import app.trackevolution.core.model.SessionDraft
import app.trackevolution.core.EventFormSessions
import app.trackevolution.core.Units
import app.trackevolution.ui.LoadState
import app.trackevolution.ui.LocalUnitSystem
import app.trackevolution.ui.TEErrorBanner
import app.trackevolution.ui.TEField
import app.trackevolution.ui.TELoadable
import app.trackevolution.ui.TESectionHeader
import app.trackevolution.ui.theme.TrackCard
import app.trackevolution.ui.theme.TrackTheme

/**
 * Create or edit an event (NS-26).
 *
 * The Car and Track fields are **free text with suggestions**, never pickers: a
 * car that isn't in the garage has to keep working, and a track nobody has been
 * to yet has to be typeable. See [EventFormModel] for why the track name is only
 * ever trimmed.
 *
 * A **new** event ends with an "Add laps" section — the event page's "Add a
 * session" card, minus the recorder: [onImport] opens the same video import,
 * whose review hands its sessions back to the form (`Route.Import.forNewEvent`),
 * and lap times can be typed straight in. Both are saved with the event, so
 * logging a day is one screen rather than a form and then a page. Editing an
 * existing event has no such section: its page already has the card.
 */
@Composable
fun EventFormScreen(
    model: EventFormModel,
    onSaved: (Int) -> Unit,
    onImport: () -> Unit = {},
    onCancel: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val colors = TrackTheme.colors

    LaunchedEffect(Unit) { if (model.state == LoadState.Loading) model.load() }
    LaunchedEffect(model.savedId) { model.savedId?.let(onSaved) }

    TELoadable(state = model.state, onRetry = model::load, modifier = modifier) {
        LazyColumn(
            modifier = Modifier.fillMaxSize().padding(horizontal = 16.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp),
            contentPadding = PaddingValues(vertical = 12.dp),
        ) {
            item("title") {
                Text(
                    if (model.editId == null) "Add event" else "Edit event",
                    style = TrackTheme.typography.h1,
                    color = colors.textStrong,
                )
            }

            item("error") { TEErrorBanner(model.error) }

            item("track") {
                SuggestingField(
                    label = "Track",
                    hint = "Pick from your tracks and known US tracks, or type a new name — " +
                        "layouts time differently, so name them separately to keep PBs honest",
                    value = model.trackName,
                    onValueChange = { model.trackName = it },
                    placeholder = "Virginia International Raceway (Full)",
                    options = model.trackOptions,
                )
            }

            item("date") {
                TEField("Start date") {
                    OutlinedTextField(
                        value = model.startDate,
                        onValueChange = { model.startDate = it },
                        placeholder = { Text("2026-05-01", style = TrackTheme.typography.sm) },
                        singleLine = true,
                        modifier = Modifier.fillMaxWidth(),
                    )
                }
            }

            item("days") {
                Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    TEField("Days", modifier = Modifier.weight(1f)) {
                        NumberField(model.days) { model.days = it }
                    }
                    TEField(
                        "On-track hours",
                        hint = "Blank uses the 2h-per-day estimate",
                        modifier = Modifier.weight(1f),
                    ) {
                        NumberField(model.trackHours, placeholder = "est. 2h per day") {
                            model.trackHours = it
                        }
                    }
                }
            }

            item("club") {
                TEField("Club / organizer") {
                    PlainField(model.club, "VIR Club") { model.club = it }
                }
            }

            item("group") {
                TEField("Run group") {
                    PlainField(model.runGroup, "High Speed") { model.runGroup = it }
                }
            }

            item("car") {
                SuggestingField(
                    label = "Car",
                    hint = "Pick from your garage or type anything",
                    value = model.car,
                    onValueChange = { model.car = it },
                    placeholder = "Corvette Z06, Miata, GT3…",
                    options = model.carOptions,
                )
            }

            item("conditions") {
                TEField("Conditions") {
                    Row(horizontalArrangement = Arrangement.spacedBy(6.dp)) {
                        ConditionChip("—", model.conditions == null) { model.conditions = null }
                        Conditions.all.forEach { option ->
                            ConditionChip(option.rawValue, model.conditions == option) {
                                model.conditions = option
                            }
                        }
                    }
                }
            }

            item("temp") {
                // Labelled and entered in the user's system; stored as whole °F
                // by the model.
                val units = LocalUnitSystem.current
                TEField("Temp ${Units.tempUnit(units)}") {
                    NumberField(model.temp, placeholder = Units.tempInputSpec(units).placeholder.toString()) {
                        model.temp = it
                    }
                }
            }

            item("best") {
                TEField(
                    "Best time",
                    hint = "Only needed when you don't log laps — logged laps compute this automatically",
                ) {
                    PlainField(model.bestTime, "2:01.24") { model.bestTime = it }
                }
            }

            item("notes") {
                TEField("Notes") {
                    OutlinedTextField(
                        value = model.notes,
                        onValueChange = { model.notes = it },
                        placeholder = {
                            Text("Weather, setup changes, incidents…", style = TrackTheme.typography.sm)
                        },
                        modifier = Modifier.fillMaxWidth(),
                    )
                }
            }

            if (model.editId == null) {
                item("laps") {
                    AddLapsCard(
                        staged = model.stagedSessions,
                        onRemoveStaged = model::removeStaged,
                        onImport = onImport,
                        label = model.sessionLabel,
                        onLabelChange = { model.sessionLabel = it },
                        laps = model.sessionLaps,
                        onLapsChange = { model.sessionLaps = it },
                        notes = model.sessionNotes,
                        onNotesChange = { model.sessionNotes = it },
                    )
                }
            }

            item("actions") {
                Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    Button(
                        onClick = model::save,
                        enabled = !model.saving && model.trackName.isNotBlank(),
                        colors = ButtonDefaults.buttonColors(
                            containerColor = colors.accent,
                            contentColor = colors.accentContrast,
                        ),
                        modifier = Modifier.testTag("eventFormSubmit"),
                    ) {
                        Text(
                            when {
                                model.editId != null -> "Save changes"
                                // The event exists and a session didn't make it:
                                // the same button now only retries the laps.
                                model.createdId != null -> "Add the laps"
                                else -> "Add event"
                            },
                            style = TrackTheme.typography.bodyStrong,
                        )
                    }
                    TextButton(onClick = onCancel) {
                        Text("Cancel", style = TrackTheme.typography.sm, color = colors.textMuted)
                    }
                }
            }
        }
    }
}

/**
 * The New Event form's "Add laps" section: sessions staged from the video
 * importer, each removable until the event is created, and a hand-typed one.
 * The copy is the event page card's, so the two read as the same thing.
 */
@Composable
private fun AddLapsCard(
    staged: List<SessionDraft>,
    onRemoveStaged: (Int) -> Unit,
    onImport: () -> Unit,
    label: String,
    onLabelChange: (String) -> Unit,
    laps: String,
    onLapsChange: (String) -> Unit,
    notes: String,
    onNotesChange: (String) -> Unit,
) {
    val colors = TrackTheme.colors
    val type = TrackTheme.typography

    TrackCard(Modifier.fillMaxWidth()) {
        TESectionHeader("Add laps")
        Text(
            "Optional — the sessions here are saved with the event. You can always add more from its page.",
            style = type.xs,
            color = colors.textMuted,
            modifier = Modifier.padding(top = 4.dp, bottom = 8.dp),
        )

        Text("Import a video", style = type.bodyStrong, color = colors.textStrong)
        Text(
            "Corvette PDR or GoPro clips already on this phone — laps, racing line and " +
                "channel graphs come out of the telemetry track. The video is read in " +
                "place, never copied or uploaded.",
            style = type.xs,
            color = colors.textMuted,
            modifier = Modifier.padding(vertical = 6.dp),
        )
        TextButton(onClick = onImport, modifier = Modifier.testTag("eventFormImportVideo")) {
            Text("Import video…", style = type.sm, color = colors.accentInk)
        }
        staged.forEachIndexed { index, draft ->
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier.fillMaxWidth().padding(top = 4.dp).testTag("stagedSession$index"),
            ) {
                Column(Modifier.weight(1f)) {
                    Text(draft.label ?: "Imported session", style = type.sm, color = colors.textStrong)
                    Text(
                        EventFormSessions.stagedSummary(draft.laps.orEmpty()),
                        style = type.xs,
                        color = colors.textMuted,
                    )
                }
                TextButton(onClick = { onRemoveStaged(index) }) {
                    Text("Remove", style = type.sm, color = colors.textMuted)
                }
            }
        }

        HorizontalDivider(
            color = colors.borderHairline,
            modifier = Modifier.padding(vertical = 14.dp),
        )

        Text("Enter lap times by hand", style = type.bodyStrong, color = colors.textStrong)
        OutlinedTextField(
            value = label,
            onValueChange = onLabelChange,
            placeholder = { Text("Day 1 — Session 2", style = type.sm) },
            singleLine = true,
            modifier = Modifier.fillMaxWidth().padding(top = 6.dp).testTag("eventFormSessionLabel"),
        )
        OutlinedTextField(
            value = laps,
            onValueChange = onLapsChange,
            placeholder = { Text("2:01.24, 2:03.1 …", style = type.sm) },
            supportingText = {
                Text(
                    "Formats: 2:01.24 · 2:01 · 121.24 (seconds)",
                    style = type.xxs,
                    color = colors.textFaint,
                )
            },
            modifier = Modifier.fillMaxWidth().padding(top = 6.dp).testTag("eventFormSessionLaps"),
        )
        OutlinedTextField(
            value = notes,
            onValueChange = onNotesChange,
            placeholder = { Text("Traffic, tire pressures, line changes…", style = type.sm) },
            modifier = Modifier.fillMaxWidth().padding(top = 6.dp),
        )
    }
}

/**
 * Free text with suggestions underneath — deliberately not a picker.
 *
 * A picker would make a track or car that isn't already in the logbook
 * unenterable, which is exactly the case a first visit is.
 */
@Composable
private fun SuggestingField(
    label: String,
    hint: String?,
    value: String,
    onValueChange: (String) -> Unit,
    placeholder: String,
    options: List<String>,
) {
    val colors = TrackTheme.colors
    var focused by remember { mutableStateOf(false) }
    val matches = remember(value, options, focused) {
        if (!focused || value.isBlank()) {
            emptyList()
        } else {
            options.filter { it.contains(value.trim(), ignoreCase = true) && !it.equals(value, true) }
                .take(5)
        }
    }

    TEField(label, hint) {
        Column {
            OutlinedTextField(
                value = value,
                onValueChange = onValueChange,
                placeholder = { Text(placeholder, style = TrackTheme.typography.sm) },
                singleLine = true,
                modifier = Modifier
                    .fillMaxWidth()
                    .onFocusChanged { focused = it.isFocused },
            )
            matches.forEach { option ->
                Text(
                    option,
                    style = TrackTheme.typography.sm,
                    color = colors.accentInk,
                    modifier = Modifier
                        .fillMaxWidth()
                        .clickable { onValueChange(option); focused = false }
                        .padding(vertical = 6.dp, horizontal = 4.dp),
                )
            }
        }
    }
}

@Composable
private fun PlainField(value: String, placeholder: String, onValueChange: (String) -> Unit) {
    OutlinedTextField(
        value = value,
        onValueChange = onValueChange,
        placeholder = { Text(placeholder, style = TrackTheme.typography.sm) },
        singleLine = true,
        modifier = Modifier.fillMaxWidth(),
    )
}

@Composable
private fun NumberField(
    value: String,
    placeholder: String = "",
    onValueChange: (String) -> Unit,
) {
    OutlinedTextField(
        value = value,
        onValueChange = onValueChange,
        placeholder = { Text(placeholder, style = TrackTheme.typography.sm) },
        singleLine = true,
        keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Decimal),
        modifier = Modifier.fillMaxWidth(),
    )
}

@Composable
private fun ConditionChip(label: String, selected: Boolean, onClick: () -> Unit) {
    val colors = TrackTheme.colors
    TrackCard(
        modifier = Modifier.clickable(onClick = onClick),
        border = if (selected) colors.accent else colors.borderHairline,
        contentPadding = 8.dp,
    ) {
        Text(
            label,
            style = TrackTheme.typography.xs,
            color = if (selected) colors.accentInk else colors.textMuted,
        )
    }
}
